### PR TITLE
[Feature] Exercise Performance Visualisation

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

This pull request introduces a new visualization for students, which they can use to compare their own performance in a course with the performance of the rest of the course. 

### Description

Using the recently introduced Participant Score table, we efficiently calculate the logged-in student's performance, the average performance of the course, and the best performance of a student for all assignments of a course. This information is then sent to the client and displayed in a Chart.js graph (see screenshot section)

Note: We only consider released and finished (assessment due date over) exercises!
### Steps for Testing

1. Log in to Artemis with a student who has some scores in a course
2. Open the course overview page for a course with some scores
3. Scroll down and you should see the exercise scores chart

### Test Coverage

exercise-scores-chart.component.ts 73% **(improving further will be very hard in a unit test as I would need to simulate a user clicking specific points on the chart, effort would be not reasonable imo)**
exercise-scores-chart.service.ts 88%
ExerciseScoresChartService.java 97%
ExerciseScoresChartResource.java 100%

### Screenshots

image

https://user-images.githubusercontent.com/29718932/114269697-bca0c200-9a08-11eb-861c-ceb80eb792ad.mov

### CODE REVIEW

**Review by willergeorg — COMMENTED** [review]
Looks great and seems to reflect the actual performances. I cannot see any discrepancies.

**Review by NicolasUllmann — COMMENTED** [review]
Tested on server 2. Everything looked accurate and good to me.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        // we only consider exercises in which the student had a chance to earn a score (released and due date over)
+        List<ExerciseScoresDTO> exerciseScoresDTOList = exerciseScoresChartService.getExerciseScores(removeNotVisibleAndUnfinishedExercise(course.getExercises()), user);
+        return ResponseEntity.ok(exerciseScoresDTOList);
+    }
+
+    private Set<Exercise> removeNotVisibleAndUnfinishedExercise(Set<Exercise> exercises) {
+        return exercises.parallelStream().filter(Exercise::isVisibleToStudents).filter(Exercise::isAssessmentDueDateOver).collect(Collectors.toSet());
+    }
+
+    public static final class Endpoints {
```
Comment (ivohashamov) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java:77]: I'm not sure if this class is necessary, but it doesn't hurt either

**Review by ivohashamov — APPROVED** [review]
Tested on ts1 and it looks good. Code is also fine from what I can see, I'm just not sure about the Endpoints nested class.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java (hunk 1 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java:77]: I found it helpful because than I can make sure the endpoints are also changed in all the integration tests :)

**Review by ArtemTsurikov — DISMISSED** [review]
Tested with @katjanakosic

**Review by katjanakosic — APPROVED** [review]
Everything works perfectly fine :)

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 2]
```diff
@@ -543,6 +543,15 @@ public void filterSensitiveInformation() {
         setGradingCriteria(null);
     }
 
+    /**
+     * Gets the type of the exercise as a string
+     *
+     * @return type of the exercise as a string
+     */
+    public String getStringRepresentationOfType() {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:551]: I don't understand why we need this?
This would be sent on all json responses of all requests (and exercises are really used very often).
However the type is already available

**Review by krusche — CHANGES_REQUESTED** [review]
please fix most of the codacy issues and delete the `getStringRepresentationOfType` methods. I think you should include this information differently in case you really need it on the client side

**Comment by stefanwaldhauser** [comment]
> please fix most of the codacy issues and delete the `getStringRepresentationOfType` methods. I think you should include this information differently in case you really need it on the client side

done :)

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java (hunk 2 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/Exercise.java]: removed

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseScoresAggregatedInformation.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        this.maxScoreAchieved = maxScoreAchieved;
+    }
+
+    public ExerciseScoresAggregatedInformation() {
+        // empty constructor for jackson
+    }
+
+    public Long getExerciseId() {
+        return exerciseId;
+    }
```
Comment (FrankeLukas) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseScoresAggregatedInformation.java:26]: Does it even make sense to offer getters for `public` variables?

File: src/main/java/de/tum/in/www1/artemis/domain/FileUploadExercise.java [hunk 4]
```diff
@@ -1,6 +1,9 @@
 package de.tum.in.www1.artemis.domain;
 
-import javax.persistence.*;
+import javax.persistence.Column;
+import javax.persistence.DiscriminatorValue;
+import javax.persistence.Entity;
+import javax.persistence.Lob;
```
Comment (FrankeLukas) [inline src/main/java/de/tum/in/www1/artemis/domain/FileUploadExercise.java:6]: The changes in this file don't seem to be related to the PR(?)

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        Course course = courseRepository.findByIdWithEagerExercisesElseThrow(courseId);
+        User user = userRepository.getUserWithGroupsAndAuthorities();
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, user);
+        // we only consider exercises in which the student had a chance to earn a score (released and due date over)
+        List<ExerciseScoresDTO> exerciseScoresDTOList = exerciseScoresChartService.getExerciseScores(removeNotVisibleAndUnfinishedExercise(course.getExercises()), user);
+        return ResponseEntity.ok(exerciseScoresDTOList);
+    }
+
+    private Set<Exercise> removeNotVisibleAndUnfinishedExercise(Set<Exercise> exercises) {
+        return exercises.parallelStream().filter(Exercise::isVisibleToStudents).filter(Exercise::isAssessmentDueDateOver).collect(Collectors.toSet());
```
Comment (FrankeLukas) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java:74]: This looks dangerous, what if an exercise has no assessment due date set (in which case `isAssessmentDueDateOver` returns `true`), but the due date of the exercise has not been reached?

File: src/main/webapp/app/course/learning-goals/learning-goal-course-detail-modal/learning-goal-course-detail-modal.component.html [hunk 6]
```diff
@@ -59,7 +59,7 @@ <h5>{{ 'artemisApp.learningGoal.learningGoalCard.connectedLectureUnits' | artemi
                             {{
                                 getLectureUnitCourseProgress(lectureUnit.id!) === undefined
                                     ? '-'
-                                    : getLectureUnitCourseProgress(lectureUnit.id!).averageScoreAchievedByStudentInLectureUnit + ' %'
+                                    : round(getLectureUnitCourseProgress(lectureUnit.id).averageScoreAchievedByStudentInLectureUnit) + ' %'
```
Comment (FrankeLukas) [inline src/main/webapp/app/course/learning-goals/learning-goal-course-detail-modal/learning-goal-course-detail-modal.component.html:62]: Doesn't removing the `!`s in this file undo the strict template fixes?

**Review by FrankeLukas — CHANGES_REQUESTED** [review]
Seems to work fine on TS2, just a few questions:

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseScoresAggregatedInformation.java (hunk 3 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/ExerciseScoresAggregatedInformation.java:26]: You are right! Made them privat b6d37968eccf8608d2e3238b65f233d5cf396252

File: src/main/java/de/tum/in/www1/artemis/domain/FileUploadExercise.java (hunk 4 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/FileUploadExercise.java:6]: Seems to be just a reformating change. Not really a problem right?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java (hunk 5 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java:74]: I do not understand what you mean? Why would this be "dangerous"? Do you have a suggestion for a fix?

File: src/main/webapp/app/course/learning-goals/learning-goal-course-detail-modal/learning-goal-course-detail-modal.component.html (hunk 6 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/learning-goals/learning-goal-course-detail-modal/learning-goal-course-detail-modal.component.html]: 1ddf78e6b95cf5a3e9154c2d6b1b45661fed7dc7 Done :) good catch!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java (hunk 5 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java]: 2cfc1951315ee9ab7c3d3a4a1402ee3aa3b03e51 Do you think this fixes the problem @FrankeLukas ?

Comment (FrankeLukas) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java]: That looks better. There would also be the special case of exercises without both due date or assessment due date, but I'm not sure if those will cause problems.

> I do not understand what you mean? Why would this be "dangerous"?

My problem was that unfinished exercises would not get filtered when they have no assessment due date. (I.e. no assessment due date, and a due date in the future => we include an exercise which is still "in progress".)

Comment (FrankeLukas) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java]: Maybe it would also be good to have a private `isExerciseFinished` method here, so that you can do `return exercises.parallelStream().filter(exercise -> isExerciseFinished(exercise)).collect(Collectors.toSet());` with something like

```
private boolean isExerciseFinished(Exercise exercise) {
    // Exercises not visible to the students are not finished either
    if (!exercise.isVisibleToStudents()) {
        return false;
    }
    
    // If an assessment due date is set, only count the exercise as finished if we passed that date
    if (exercise.getAssessmentDueDate() != null) {
        return ZonedDateTime.now().isAfter(exercise.getAssessmentDueDate());
    }
    
    // If no assessment due date is set, make sure to only count exercises which have a passed due date
    return exercise.getDueDate() != null && ZonedDateTime.now().isAfter(exercise.getDueDate());
}
```

Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseScoresChartResource.java]: Very nice! Looks much cleaner. Did it in e93dbd91abae7cc727edfe2c2d98a4fa5a9234b4

File: src/main/webapp/app/shared/util/utils.ts [hunk 7]
```diff
@@ -106,3 +106,21 @@ export const checkForMissingTranslationKey = (alert: JhiAlert) => {
...
+            output.push(word[i].toUpperCase());
+        } else {
+            if (i > 0 && regex.test(word[i])) {
+                output.push(' ');
+            }
+            output.push(word[i]);
+        }
+    }
+    return output.join('');
+};
```
Comment (ivanchimeno) [inline src/main/webapp/app/shared/util/utils.ts:126]: Well this looks like magic 😆

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Tried it on TS1: Maybe this is an issue because I only have 1 exrcise, but  I guess those edge cases should be fixed:
image

**Review by ArtemTsurikov — APPROVED** [review]
Tested with @katjanakosic. Everything works.

**Review by maximilianbergmann — APPROVED** [review]
Looks good. Just a suggestion, I would maybe scale the graphs to the width of the statistics above or to 100%.

**Comment by stefanwaldhauser** [comment]
> Tried it on TS1: Maybe this is an issue because I only have 1 exrcise, but I guess there is some extra work needed on the displaying.
> image

Can you please send me the course? On what browser is this?

**Comment by stefanwaldhauser** [comment]
> Tried it on TS1: Maybe this is an issue because I only have 1 exrcise, but I guess there is some extra work needed on the displaying.
> image

Fixed now in d187989098e7a007545362988ace95b8d1a742a8. Please check again

**Review by maximilianbergmann — CHANGES_REQUESTED** [review]
Looks good on large screens. Unfortunately, it's not right on small screens / smartphones. The statistic exceeds the normal width.

Artemis
Artemis1

**Comment by stefanwaldhauser** [comment]
> Looks good on large screens. Unfortunately, it's not right on small screens / smartphones. The statistic exceeds the normal width.
> 
> Artemis
> Artemis1

Honestly I do not know how to fix this. Anybody got an idea?

**Review by NicolasUllmann — COMMENTED** [review]
Tested on server 1. As mentioned above on smartphone screens, there seems to be a problem with the visualization. But on a 13' Laptop screen everything works fine.

**Review by willergeorg — COMMENTED** [review]
Tested on TS2. Seems good for me on large screens.

**Review by cremertim — COMMENTED** [review]
Tested on ts1 - everything workout ot for me on a large screen

**Comment by stefanwaldhauser** [comment]
The problem with the small screens should be fixed now. @iozbeyli  thanks for the css help :)

**Review by maximilianbergmann — DISMISSED** [review]
Tested again. (TS3) Looks great now on every screens :-)

**Review by ivohashamov — APPROVED** [review]
Changes look good on TS2, I tried making my browser window as small as possible and I think that scenario was handled pretty well. Since there's a minimum width limit, I'm not sure how the page would be displayed on a phone, but that also doesn't seem very relevant. Nice job!

File: src/main/java/de/tum/in/www1/artemis/domain/FileUploadExercise.java (hunk 4 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/domain/FileUploadExercise.java:6]: It is a problem, please avoid it! It's not good if PRs randomly change the imports due to different IntelliJ settings.
@MaisiKoleni recently added some configs that should be synced directly with IntelliJ.
Please make sure to have the correct project settings in IntelliJ to avoid it.

File: src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java [hunk 8]
```diff
@@ -16,7 +16,10 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.fasterxml.jackson.annotation.*;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:22]: avoid changing imports

Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java]: Done: 60bb32b6b5fd63244a569bab1586bc7405354b30

**Review by ivohashamov — APPROVED** [review]
Reapprove

**Review by SimonEntholzer — APPROVED** [review]
Tested on TS2, now everything works fine :D

**Review by coolchock — APPROVED** [review]
tested on ts2, works as expected
